### PR TITLE
Fix(dplan 1965): improve error validation of the Datepicker

### PR DESCRIPTION
**Pull request by sakutademos**
**Ticket:** [DPLAN-1965](https://demoseurope.youtrack.cloud/issue/DPLAN-1965/Invalid-Inputs-beim-Verfahrenerstellen-werfen-nicht-die-richtigen-Fehlernachrichten) 

**Description:**  This PR adjusts the validation of the Datepicker.

- add the `data-dp-validate-error-fieldname` attribute to the Datepicker via `setAttribute() `to facilitate the display of the field name in case of an error.
- include `name` attributes `r_startdate` and `r_enddate` in the validation form, as the date field need to be validated every time the value changes and not only if the field has an `required` attribute
- an empty string is accepted if the `required` attribute is not present.

[PR in Core](https://github.com/demos-europe/demosplan-core/pull/3038)

### CODE REVIEW

File: src/lib/validation/utils/validateDatepicker.js [hunk 1]
```diff
@@ -2,11 +2,19 @@ import { toggleErrorClass } from './helpers'
 
 export default function validateDatepicker (input) {
   let isValid = true
+
   if (input.value === '') {
-    isValid = false
+    isValid = !input.hasAttribute('required')
```
Comment (spiess-demos) [inline src/lib/validation/utils/validateDatepicker.js:7]: I don't understand this logic. for me this means, setting isValid to true for all instances where no input is required. But, i can also input invalid data into a non-required field!

File: src/lib/validation/utils/helpers.js [hunk 2]
```diff
@@ -2,7 +2,7 @@ import getScrollTop from '../../../utils/getScrollTop'
 import prefixClass from '../../../utils/prefixClass'
 
 function getAllInputsArray (form) {
-  return Array.from(form.querySelectorAll('[pattern], [required], input[type="email"], .tiptap__input--hidden.is-required, .multiselect.is-required, fieldset.is-required, [minlength], [maxlength], [data-dp-validate-maxlength], [data-dp-validate-should-equal]'))
+  return Array.from(form.querySelectorAll('[pattern], [required], input[type="email"], input[name="r_startdate"], input[name="r_enddate"], .tiptap__input--hidden.is-required, .multiselect.is-required, fieldset.is-required, [minlength], [maxlength], [data-dp-validate-maxlength], [data-dp-validate-should-equal]'))
```
Comment (spiess-demos) [inline src/lib/validation/utils/helpers.js:5]: I'd opt against hardcoding business logic from demosplan-core here. Instead, we could always validate date inputs with ```[data-ad-id]``` (which is produced by the datepicker lib).

File: src/lib/validation/utils/helpers.js [hunk 3]
```diff
@@ -16,6 +16,8 @@ function shouldValidate (input) {
     isRequired ||
     input.hasAttribute('pattern') ||
     input.getAttribute('type') === 'email' ||
+    input.getAttribute('name') === 'r_startdate' ||
+    input.getAttribute('name') === 'r_enddate' ||
```
Comment (spiess-demos) [inline src/lib/validation/utils/helpers.js:20]: see the other comment. let's change this into:

```suggestion
    input.hasAttribute('data-ad-id')
```

and always validate date fields.

File: src/lib/validation/utils/validateDatepicker.js [hunk 4]
```diff
@@ -2,11 +2,19 @@ import { toggleErrorClass } from './helpers'
...
-    const regex = /^(?:(?:31(.)(?:0?[13578]|1[02]))\1|(?:(?:29|30)(.)(?:0?[13-9]|1[0-2])\2))(?:(?:1[6-9]|[2-9]\d)?\d{2})$|^(?:29(.)0?2\3(?:(?:(?:1[6-9]|[2-9]\d)?(?:0[48]|[2468][048]|[13579][26])|(?:(?:16|[2468][048]|[3579][26])00))))$|^(?:0?[1-9]|1\d|2[0-8])(.)(?:(?:0?[1-9])|(?:1[0-2]))\4(?:(?:1[6-9]|[2-9]\d)?\d{2})$/
+    /**
+     * DpDatePicker accepts the dateFormat: 'dd.mm.yyyy'
+     * Regex to check date from https://www.regextester.com/97612
+     * Allows dates following these rules:
+     * - days may have leading zeros. 1-31. max 2 digits
+     * - months may have leading zeros. 1-12. max 2 digits
+     * - years 1900-2099. 4 digits
+     */
+    const regex = /^\s*(3[01]|[12][0-9]|0?[1-9])\.(1[012]|0?[1-9])\.((?:19|20)\d{2})\s*$/
```
Comment (spiess-demos) [inline src/lib/validation/utils/validateDatepicker.js:17]: While narrowing down the possible input to meet the Datepicker capabilities is fine (reads: no blocker here), i feel like we should possibly allow users to input dates as they like.

Another question: how do we educate users about the expected date format? the datepicker lib does not offer a placeholder, so...

**Review by spiess-demos — DISMISSED** [review]

File: src/lib/validation/utils/validateDatepicker.js (hunk 1 above)
Comment (sakutademos) [reply src/lib/validation/utils/validateDatepicker.js:7]: it is about the empty string, whether the empty string is allowed or not. The problem is that we have to validate the Datepicker every time the user enters smth, not only if this field is required. Similar to the email field, it should be validated every time because there is a pattern. In this case, we can allow an empty string, like [here](https://github.com/demos-europe/demosplan-ui/blob/main/src/lib/validation/utils/validateInputField.js#L36)

File: src/lib/validation/utils/helpers.js (hunk 2 above)
Comment (sakutademos) [reply src/lib/validation/utils/helpers.js]: Hmm, if this is a specific Datepicker attribute, then we can do it. Thanks

File: src/lib/validation/utils/helpers.js (hunk 3 above)
Comment (sakutademos) [reply src/lib/validation/utils/helpers.js]: done

File: src/components/DpDatepicker/DpDatepicker.vue [hunk 5]
```diff
@@ -142,12 +148,22 @@ export default {
   },
 
   methods: {
+    addErrorFieldnameAttribute () {
+      const datePickerInput = document.getElementsByName(this.name)[0]
+      /**
+       * This attribute is needed for validation to display the field name in case of an error
+       * and must be set every time the Datepicker is mounted or updated.
+       */
```
Comment (salisdemos) [inline src/components/DpDatepicker/DpDatepicker.vue:156]: why does it need an update on every change?

File: src/lib/validation/utils/helpers.js [hunk 6]
```diff
@@ -16,6 +16,7 @@ function shouldValidate (input) {
     isRequired ||
     input.hasAttribute('pattern') ||
     input.getAttribute('type') === 'email' ||
+    input.hasAttribute('data-ad-id') ||
```
Comment (salisdemos) [inline src/lib/validation/utils/helpers.js:19]: where comes this from? Why is this needed?

File: src/lib/validation/utils/validateDatepicker.js (hunk 4 above)
Comment (salisdemos) [reply src/lib/validation/utils/validateDatepicker.js:17]: A solution could be to reformat the date before passing it to the datepicker. But that would be a story by its own.

File: src/lib/validation/utils/helpers.js (hunk 6 above)
Comment (sakutademos) [reply src/lib/validation/utils/helpers.js]: the Datepicker uses this specific [attribute](https://github.com/mathislucka/a11y-datepicker/blob/master/src/js/draw/Draw.js#L27). See here a comment from @spiess-demos https://github.com/demos-europe/demosplan-ui/pull/841#discussion_r1587279898 . I think it is a good idea to get the Datepicker input with this attribute, but I think I should add a comment for a better readability

File: src/components/DpDatepicker/DpDatepicker.vue (hunk 5 above)
Comment (sakutademos) [reply src/components/DpDatepicker/DpDatepicker.vue:156]: after updating the input value, the `data-dp-validate-error-fieldname` attribute that was added during mounting disappears from DOM, so I also [include it in the `emitUpdate` action](https://github.com/demos-europe/demosplan-ui/pull/841/files#diff-9749bd9f6c18eb9c907ecf005c35b25dc64bc5431f99f694aba87c0ee5c1c519R166)

Comment (salisdemos) [reply src/components/DpDatepicker/DpDatepicker.vue:156]: Thanks for the explanation

File: src/lib/validation/utils/helpers.js (hunk 6 above)
Comment (salisdemos) [reply src/lib/validation/utils/helpers.js]: Ok. Thanks. The Comment helps a lot

**Review by salisdemos — APPROVED** [review]

File: src/lib/validation/utils/validateDatepicker.js (hunk 4 above)
Comment (spiess-demos) [reply src/lib/validation/utils/validateDatepicker.js:17]: thanks for adding the tooltip. for now this is ok. in the long run we should display the expected format permanently, as it could easily missed in a tooltip.
